README.md
# IntisTelecomCandidateTest

[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Quality Score][ico-code-quality]][link-code-quality]

1. Разработать скрипт , принимающий курсы валют из двух источников и обновляющий данные в таблице mysql. Без использования каких либо фреймворков.

Таблица:
--------
`currency`
- `symbol` (код валюты)
- `rate` (курс валюты)

Источники:
----------
- `http://localhost/rates1.json`
```
{ "rates" : [{ "symbol": "USD", "rate":1 }, { "symbol": "EUR", "rate":2 }, { "symbol": "RUR", "rate":3 }] }
```
- Локальный файл rates.json, содержащий
```
[ { "USD": 1 }, { "EUR": 2 }, { "RUR": 3 } ]
```
- источники необходимо обеспечить самостоятельно
- выбор варианта обновления должен осуществляться параметром при запуске скрипта из консоли (например php update.php 1/2, где 1 или 2 - выбор варианта обновления)

2. Написать юнит-тесты для разработанного кода

## Install

Via Composer

``` bash
$ composer require iGusev/IntisTelecomCandidateTest
```

## Usage

``` php
$skeleton = new League\Skeleton();
echo $skeleton->echoPhrase('Hello, League!');
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email mail@igusev.ru instead of using the issue tracker.

## Credits

- [Ilya Gusev][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/iGusev/IntisTelecomCandidateTest.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/iGusev/IntisTelecomCandidateTest/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/iGusev/IntisTelecomCandidateTest.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/iGusev/IntisTelecomCandidateTest.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/iGusev/IntisTelecomCandidateTest.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/iGusev/IntisTelecomCandidateTest
[link-travis]: https://travis-ci.org/iGusev/IntisTelecomCandidateTest
[link-scrutinizer]: https://scrutinizer-ci.com/g/iGusev/IntisTelecomCandidateTest/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/iGusev/IntisTelecomCandidateTest
[link-downloads]: https://packagist.org/packages/iGusev/IntisTelecomCandidateTest
[link-author]: https://github.com/:author_username
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `IntisTelecomCandidateTest` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ExampleTest.php
<?php

namespace iGusev\IntisTelecomCandidateTest;

class ExampleTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


tests/Command/ImportCurrencyCommandTest.php
<?php
/**
 * Created by PhpStorm.
 * User: iGusev
 * Date: 02/06/16
 * Time: 02:07
 */

namespace Command;

use iGusev\IntisTelecomCandidateTest\Command\ImportCurrencyCommand;
use iGusev\IntisTelecomCandidateTest\InvalidJSONException;
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Tester\CommandTester;
use \Mockery;

class ImportCurrencyCommandTest extends \PHPUnit_Framework_TestCase
{

    const SUCCESS_IMPORT = '/Импорт успешно завершен/';
    const PARAMS_NOT_FOUND = '/Отсутствуют обязательные параметры/';
    const FILE_NOT_FOUND = '/Файл (.*) недоступен/';
    const INVALID_JSON = '/Невалидный JSON в файле/';

    /**
     * @var Mockery\MockInterface|Command
     */
    protected $command;
    /**
     * @var CommandTester
     */
    protected $commandTester;

    public function incorrectFilePathProvider()
    {
        return [
            ['/123.json']
        ];
    }

    public function incorrectUrlPathProvider()
    {
        return [
            ['https://google.com/123.json']
        ];
    }

    public function correctFilePathProvider()
    {
        return [
            ['tests/fixtures/rates.json']
        ];
    }

    public function correctUrlPathProvider()
    {
        return [
            ['tests/fixtures/url-rates.json']
        ];
    }

    public function setUp()
    {
        $mock = Mockery::mock(ImportCurrencyCommand::class.'[setData]', [null])->shouldAllowMockingProtectedMethods();

        $application = new Application();
        $application->add($mock);

        $this->command = $application->find('import:currency');
        $this->commandTester = new CommandTester($this->command);
    }

    public function tearDown()
    {
        unset($this->command);
        unset($this->commandTester);

        Mockery::close();
    }

    public function testExecuteWithoutArguments()
    {
        $this->commandTester->execute([
            'command' => $this->command->getName()
        ]);

        $this->assertRegExp(static::PARAMS_NOT_FOUND, $this->commandTester->getDisplay());
    }

    /**
     * @param $incorrectFilePath
     *
     * @dataProvider incorrectFilePathProvider
     */
    public function testExecuteFileNotFoundFile($incorrectFilePath)
    {
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--file' => $incorrectFilePath
        ]);

        $this->assertRegExp(static::FILE_NOT_FOUND, $this->commandTester->getDisplay());
    }

    /**
     * @param $incorrectUrlPath
     *
     * @dataProvider incorrectUrlPathProvider
     */
    public function testExecuteFileNotFoundUrl($incorrectUrlPath)
    {
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--url' => $incorrectUrlPath
        ]);

        $this->assertRegExp(static::FILE_NOT_FOUND, $this->commandTester->getDisplay());
    }

    /**
     * @param $correctFilePath
     *
     * @dataProvider correctFilePathProvider
     */
    public function testExecuteWithFile($correctFilePath)
    {
        $this->command->shouldDeferMissing('setData')->shouldReceive('setData')->andReturn(true);
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--file' => realpath($correctFilePath)
        ]);

        $this->assertRegExp(static::SUCCESS_IMPORT, $this->commandTester->getDisplay());
    }

    /**
     * @param $correctFilePath
     *
     * @dataProvider correctUrlPathProvider
     */
    public function testExecuteWithUrl($correctFilePath)
    {
        $this->command->shouldDeferMissing('setData')->shouldReceive('setData')->andReturn(true);
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--url' => realpath($correctFilePath)
        ]);

        $this->assertRegExp(static::SUCCESS_IMPORT, $this->commandTester->getDisplay());
    }

    public function testInvalidJsonException1()
    {
        $this->command->shouldDeferMissing('setData')->shouldReceive('setData')->andReturn(true);
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--file' => realpath('tests/fixtures/invalid-rates.json')
        ]);

        $this->assertRegExp(static::INVALID_JSON, $this->commandTester->getDisplay());
    }

    public function testInvalidJsonException2()
    {
        $this->command->shouldDeferMissing('setData')->shouldReceive('setData')->andReturn(true);
        $this->commandTester->execute([
            'command' => $this->command->getName(),
            '--url' => realpath('tests/fixtures/invalid-rates.json')
        ]);

        $this->assertRegExp(static::INVALID_JSON, $this->commandTester->getDisplay());
    }


}


src/Command/ImportCurrencyCommand.php
<?php
/**
 * Created by PhpStorm.
 * User: iGusev
 * Date: 02/06/16
 * Time: 02:05
 */

namespace iGusev\IntisTelecomCandidateTest\Command;

use iGusev\IntisTelecomCandidateTest\BaseException;
use iGusev\IntisTelecomCandidateTest\FileNotFoundException;
use iGusev\IntisTelecomCandidateTest\InvalidJSONException;
use Illuminate\Database\Capsule\Manager;
use Illuminate\Database\Connection;
use JsonSchema\RefResolver;
use JsonSchema\Uri\UriResolver;
use JsonSchema\Uri\UriRetriever;
use JsonSchema\Validator;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * Команда импорта валют
 *
 * Class ImportCurrencyCommand
 * @package iGusev\IntisTelecomCandidateTest\Command
 */
class ImportCurrencyCommand extends Command
{
    /**
     * Таблица с валютами
     */
    const TABLE_CURRENCY = 'currency';

    /**
     * Путь к схеме импорта из файлов
     */
    const FILE_SCHEMA = 'src/config/file-schema.json';

    /**
     * Путь к схеме импорта из урлов
     */
    const URL_SCHEMA = 'src/config/url-schema.json';

    /**
     * @var Connection
     */
    protected $connection;

    /**
     * Инициализация команды
     */
    protected function configure()
    {
        $this->setName('import:currency')
            ->setDescription('Импорт валют из различных источников')
            ->addOption(
                'file',
                'f',
                InputOption::VALUE_REQUIRED,
                'Если установлен, будет произведен ' .
                'импорт из указанного файла'
            )
            ->addOption(
                'url',
                'u',
                InputOption::VALUE_REQUIRED,
                'Если установлен, будет произведен ' .
                'импорт из указанного url-адреса'
            );

        $manager = new Manager();

        $manager->addConnection([
            'driver' => 'mysql',
            'host' => 'localhost',
            'database' => 'intis_candidate_test',
            'username' => 'root',
            'password' => 'rootroot',
            'charset' => 'utf8',
            'collation' => 'utf8_unicode_ci',
            'prefix' => '',
        ]);

        $this->connection = $manager->getConnection();
    }

    /**
     * Основная логика
     *
     * @param InputInterface $input
     * @param OutputInterface $output
     *
     * @return string
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        try {
            if ($input->getOption('file')) {
                $this->importFromFile($input->getOption('file'));
                $output->writeln($input->getOption('file'));
            }
            if ($input->getOption('url')) {
                $this->importFromUrl($input->getOption('url'));
                $output->writeln($input->getOption('url'));
            }

            if (!$input->getOption('file') && !$input->getOption('url')) {
                $output->writeln("<question>Отсутствуют обязательные параметры. " .
                    "Для вызова справки используйте:\n" .
                    "bin/console help import:currency</question>");
            } else {
                $output->writeln('<info>Импорт успешно завершен</info>');
            }
        } catch (BaseException $e) {
            $output->writeln("<error>{$e->getMessage()}</error>");
        }
    }

    /**
     * Импорт из файла
     *
     * @param string $filepath
     *
     * @return bool
     *
     * @throws InvalidJSONException
     */
    protected function importFromFile(string $filepath): bool
    {
        $data = json_decode($this->getData($filepath));

        if ($this->validateJson($data, static::FILE_SCHEMA)) {
            foreach ($data as $item) {
                $this->setData(key($item), reset($item));
            }

            return true;
        }

        throw new InvalidJSONException("Невалидный JSON в файле {$filepath}");
    }

    /**
     * Импорт по урлу
     *
     * @param string $url
     *
     * @return bool
     *
     * @throws InvalidJSONException
     */
    protected function importFromUrl(string $url): bool
    {
        $data = json_decode($this->getData($url));

        if ($this->validateJson($data, static::URL_SCHEMA)) {
            foreach ($data->rates as $item) {
                $this->setData($item->symbol, $item->rate);
            }

            return true;
        }

        throw new InvalidJSONException("Невалидный JSON в файле {$url}");
    }

    /**
     * Получение содержимого
     *
     * @param string $path
     *
     * @return string
     *
     * @throws FileNotFoundException
     */
    protected function getData(string $path)
    {
        if (($data = @file_get_contents($path)) !== false) {
            return $data;
        }

        throw new FileNotFoundException("Файл {$path} недоступен");
    }

    /**
     * Запись в бд
     *
     * @param string $symbol
     * @param int $rate
     *
     * @return bool
     *
     * @codeCoverageIgnore
     */
    protected function setData(string $symbol, int $rate): bool
    {
        return $this->connection->table(static::TABLE_CURRENCY)
            ->updateOrInsert(['symbol' => $symbol], ['rate' => $rate]);
    }

    /**
     * Валидация файлов по json-схеме
     *
     * @param $data
     * @param string $schemaPath
     *
     * @return bool
     */
    protected function validateJson($data, string $schemaPath)
    {
        $refResolver = new RefResolver(new UriRetriever(), new UriResolver());
        $schema = $refResolver->resolve('file://' . realpath($schemaPath));

        $validator = new Validator();
        $validator->check($data, $schema);

        return $validator->isValid();
    }
}
